### ISSUE TITLE
Theano MaxPooling1D/MaxPooling2D is off-by-one

### ISSUE THREAD

**Issue by nneonneo** [fchollet/keras#6936]
MaxPooling1D and MaxPooling2D are off-by-one with the Theano backend when using padding='same'.

I have tested with both the latest master Theano and 0.9.0 release Theano, and with the latest master Keras and 2.0.4 release Keras.

Here's a simple test script:

```
import keras
import numpy as np
from keras.models import Sequential
from keras.layers import MaxPooling1D

model = Sequential()
model.add(MaxPooling1D(2, input_shape=(4, 1), padding='same'))
model.compile(loss='categorical_crossentropy', optimizer='rmsprop')

inarr = np.array([1,2,3,4])

print(model.predict(inarr.reshape((1,4,1))))
```

With Tensorflow this outputs
```
Using TensorFlow backend.
[[[ 2.]
  [ 4.]]]
```

as expected. But with Theano this outputs

```
Using Theano backend.
[[[ 1.]
  [ 3.]]]
```

which is incorrect - it appears to pad on the left of the signal and neglect the last input value (even though the input size is divisible by the pool size).

UPDATE: Just tested MaxPooling2D - same issue with Theano (unsurprisingly).

**Comment by nouiz** [comment]
This is a keras issue. Maybe/probably it build the wrong Theano graph.

If you convert the keras code to pure Theano, I can look at it. But otherwise I won't have time in the short term.

If you have an issue on keras about that, link to it here.

**Comment by nneonneo** [comment]
I believe this is filed as a Keras bug not a Theano bug. I think the Keras Theano backend is generating bad code, as you say.
